File: examples/zig/stateful-counter/Counter_test.zig
const std = @import("std");

const root = @import("../examples_test.zig");
const Counter = @import("Counter.runar.zig").Counter;

const contract_source = @embedFile("Counter.runar.zig");

test "compile-check Counter.runar.zig" {
    const allocator = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    try tmp.dir.writeFile(.{ .sub_path = "Counter.runar.zig", .data = contract_source });

    const path = try std.fs.path.join(allocator, &.{ ".zig-cache", "tmp", tmp.sub_path[0..], "Counter.runar.zig" });
    defer allocator.free(path);
    try root.runar.compileCheckFile(allocator, path);
    try root.runar.compileCheckSource(allocator, contract_source, "Counter.runar.zig");
}

test "counter executes increment and decrement directly" {
    var counter = Counter.init(2);

    counter.increment();
    try std.testing.expectEqual(@as(i64, 3), counter.count);

    counter.decrement();
    try std.testing.expectEqual(@as(i64, 2), counter.count);
}

test "counter decrement at zero fails through the real contract assertion path" {
    try root.runar.expectAssertFailure(std.testing.allocator, "counter-underflow");
}

File: examples/go/stateful-counter/Counter_test.go
package contract

import (
	"testing"
	runar "github.com/icellan/runar/packages/runar-go"
)

func TestCounter_Increment(t *testing.T) {
	c := &Counter{Count: 0}
	c.Increment()
	if c.Count != 1 {
		t.Errorf("expected Count=1, got %d", c.Count)
	}
}

func TestCounter_IncrementMultiple(t *testing.T) {
	c := &Counter{Count: 0}
	for i := 0; i < 10; i++ {
		c.Increment()
	}
	if c.Count != 10 {
		t.Errorf("expected Count=10, got %d", c.Count)
	}
}

func TestCounter_Decrement(t *testing.T) {
	c := &Counter{Count: 5}
	c.Decrement()
	if c.Count != 4 {
		t.Errorf("expected Count=4, got %d", c.Count)
	}
}

func TestCounter_DecrementAtZero_Fails(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected assertion failure")
		}
	}()
	c := &Counter{Count: 0}
	c.Decrement()
}

func TestCounter_IncrementThenDecrement(t *testing.T) {
	c := &Counter{Count: 0}
	c.Increment()
	c.Increment()
	c.Increment()
	c.Decrement()
	if c.Count != 2 {
		t.Errorf("expected Count=2, got %d", c.Count)
	}
}

func TestCounter_Compile(t *testing.T) {
	if err := runar.CompileCheck("Counter.runar.go"); err != nil {
		t.Fatalf("Rúnar compile check failed: %v", err)
	}
}

// Row 489: Counter initial state is count=0
func TestCounter_InitialState(t *testing.T) {
	c := &Counter{Count: 0}
	if c.Count != 0 {
		t.Errorf("expected initial Count=0, got %d", c.Count)
	}
}
